### PR TITLE
Fix polymorphic compare on JavaScript values

### LINKED ISSUE #895: [BUG] JS string comparison raises an exception
**Describe the bug**
I've noticed that code like `Dom_html.document##.readyState = Js.string "loading"` throws an exception. Exception has the following message: `compare: abstract value`.
Here is the stacktrace from my production code:

```caml_raise_with_arg | @ | stdlib.js:117
  | caml_raise_with_string | @ | stdlib.js:122
  | caml_invalid_argument | @ | stdlib.js:178
  | caml_compare_val | @ | stdlib.js:424
  | caml_equal | @ | stdlib.js:455
  | domContentLoaded | @ | lwt_js_events.ml:537
```
It seems that the code fails because of the `caml_compare_val` function, in the following [line](https://github.com/ocsigen/js_of_ocaml/blob/0052b39e079a64d6886a29c4e66a41b4ea9e63c0/runtime/stdlib.js#L424).
The `Lwt_js_events.domContentLoaded` function which calls `caml_equal` just compares `Dom_html.document##.readyState` with some constant at line 537.

Here is a another simple test illustrating the issue:
```ocaml
(* This is ok and shows [true] in console *)
print_endline "comparing two equal js strings";
print_endline (string_of_bool (Js.string "foo" = Js.string "foo"));
(* This is ok and shows [false] in console *)
print_endline "comparing two different js strings";
print_endline (string_of_bool (Js.string "loading" = Js.string "complete"));
(* This one fails*)
let state = Dom_html.document##.readyState in
print_endline "comparing state";
print_endline (string_of_bool (state = Js.string "complete"))
```

**Expected behavior**
This should not fail.

**Versions**
jsoo commit 0052b39e079a64d6886a29c4e66a41b4ea9e63c0

### PR DESCRIPTION
**Pull request by hhugo**
- [x] Compare works on javascript `string` and `String` object
- [x] Compare works on javascript `number` and `Number` object

### CODE REVIEW

File: runtime/stdlib.js [hunk 1]
```diff
@@ -406,7 +406,18 @@ function caml_compare_val (a, b, total) {
       } else if (b instanceof MlBytes ||
                  (b instanceof Array && b[0] === (b[0]|0))) {
         return -1;
-      } else if (typeof a != "number") {
+      } else if ((typeof a === "number" || typeof a === "boolean" || typeof a === "string") &&
```
Comment (TyOverby) [inline runtime/stdlib.js]: This would permit `a` being a `number` and `b` being a `string`.  Is this something that we have to worry about?

**Comment by hhugo** [comment]
@vouillon, another pair of eyes on the diff of `stdlib.js` would be nice. https://github.com/ocsigen/js_of_ocaml/pull/896/files#diff-1c1ab2d06ff97085527fbab9c3084e42

File: runtime/stdlib.js (hunk 1 above)
Comment (hhugo) [reply runtime/stdlib.js]: We need to consider every possible combination of types. I rewrote the function to fix the code / make the code clearer.

**Comment by hhugo** [comment]
cc @xclerc

File: runtime/stdlib.js [hunk 2]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
+          break
+        case 250: // Forward tag
+          // Cannot happen, handled above
+          caml_invalid_argument("equal: got Forward_tag, should not happen");
+          break;
+        case 251: //Abstract
+          caml_invalid_argument("equal: abstract value");
+          break;
+        case 252: // OCaml string
           if (a !== b) {
```
Comment (hhugo) [inline runtime/stdlib.js]: This is already checked above

File: lib/js_of_ocaml/dom.ml [hunk 3]
```diff
@@ -263,7 +264,7 @@ let nodeType e =
 
 module CoerceTo = struct
   let cast (e : #node Js.t) t =
-    if e##.nodeType = t then Js.some (Js.Unsafe.coerce e) else Js.null
+    if e##.nodeType == t then Js.some (Js.Unsafe.coerce e) else Js.null
```
Comment (TyOverby) [inline lib/js_of_ocaml/dom.ml]:
```suggestion
    if phys_equal e##.nodeType t then Js.some (Js.Unsafe.coerce e) else Js.null
```

File: lib/js_of_ocaml/import.ml [hunk 4]
```diff
@@ -0,0 +1,71 @@
+(* Js_of_ocaml
```
Comment (TyOverby) [inline lib/js_of_ocaml/import.ml]: Is there a place we could put this Import module where it could be shared?

File: lib/lwt/lwt_xmlHttpRequest.ml [hunk 5]
```diff
@@ -166,7 +167,7 @@ let perform_raw
     | Some (`String _ | `Blob _) -> override_method "POST", content_type
   in
   let url =
-    if get_args = []
+    if Poly.( = ) get_args []
```
Comment (TyOverby) [inline lib/lwt/lwt_xmlHttpRequest.ml]:
```suggestion
    if Poly.(get_args = [])
```

File: runtime/stdlib.js [hunk 6]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
   var stack = [];
   for(;;) {
     if (!(total && a === b)) {
-      if (a instanceof MlBytes) {
-        if (b instanceof MlBytes) {
+      // Compute TAG for [a]
+      var ta = 1001;
```
Comment (TyOverby) [inline runtime/stdlib.js]: What does "ta" stand for?

File: runtime/stdlib.js [hunk 7]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
   var stack = [];
   for(;;) {
     if (!(total && a === b)) {
```
Comment (TyOverby) [inline runtime/stdlib.js]: What is "total"?

File: runtime/stdlib.js [hunk 8]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
-      if (a instanceof MlBytes) {
-        if (b instanceof MlBytes) {
+      // Compute TAG for [a]
+      var ta = 1001;
+      if (typeof a === "number") ta = 1000;
```
Comment (TyOverby) [inline runtime/stdlib.js]: The magic numbers in this code make it really hard to follow.  Could these be pulled out into variables?

File: runtime/stdlib.js [hunk 9]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
-            if (cmp != 0) return cmp;
+          if(!comp)
+            caml_invalid_argument("compare: abstract value");
+          var x = comp(a,b,total);
+          if (x != 0) return x;
+          break;
+        case 1256: // compare function
+          var x = a.compare(b,total);
+          if(x != x | 0)
+            caml_invalid_argument("compare: abstract value");
```
Comment (TyOverby) [inline runtime/stdlib.js]: What scenarios trigger this code path?

File: runtime/stdlib.js [hunk 10]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
+          break;
+        case 1000: // Number
+        case 1001: // The rest
+          // Here we can be in the following cases:
+          // - a and b are primitive types, comparison works as expected.
+          // - a or b aren't primitive types, comparison always return false.
+          if (a < b) return -1;
+          if (a > b) return 1;
+          if (a != b) {
+            if (!total) return NaN;
```
Comment (TyOverby) [inline runtime/stdlib.js]: Returning NaN here is very surprising.  What does this mean? Is there a native OCaml equivalent?

Comment (TyOverby) [reply runtime/stdlib.js]: Also, what is "total"

File: runtime/stdlib.js [hunk 11]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
+          if (x != 0) return x;
+          break;
+        case 1256: // compare function
+          var x = a.compare(b,total);
+          if(x != x | 0)
+            caml_invalid_argument("compare: abstract value");
+          x = x | 0;
+          if (x != 0) return x;
+          break;
+        case 1000: // Number
```
Comment (TyOverby) [inline runtime/stdlib.js]: What happens when integers are compared against floats?

File: runtime/stdlib.js [hunk 12]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
...
-        if (a != b) {
-          if (!total) return NaN;
-          if (a == a) return 1;
-          if (b == b) return -1;
+          break;
+        case 246: // Lazy_tag
+        case 254: // Double_array
+        default: // Block with other tag
+          if (a.length != b.length) return (a.length < b.length)?-1:1;
+          if (a.length > 1) stack.push(a, b, 1);
```
Comment (TyOverby) [inline runtime/stdlib.js]: Is this not just pushing the same objects back on the stack?

File: runtime/stdlib.js [hunk 13]
```diff
@@ -352,85 +352,129 @@ function caml_compare_val (a, b, total) {
   var stack = [];
```
Comment (TyOverby) [inline runtime/stdlib.js]: It looks like this stack is only pushed onto.  Is that correct?

File: runtime/stdlib.js (hunk 6 above)
Comment (hhugo) [reply runtime/stdlib.js]: Tag of a

File: runtime/stdlib.js (hunk 9 above)
Comment (hhugo) [reply runtime/stdlib.js]: Any javascript value with a compare method returning non integer result

File: runtime/stdlib.js (hunk 7 above)
Comment (hhugo) [reply runtime/stdlib.js]: `total` say whether we need a total order or not.
We only need a total order for polymorphic compare, not for boolean operator  `=`, `<>`, `<`, `>`, ...

File: runtime/stdlib.js (hunk 12 above)
Comment (hhugo) [reply runtime/stdlib.js]: The stack looks like `[offset, objectleft, objectright, offset, objectleft, objectright]`.
Here we push the same objects with offset `1`. The next time we pick elements from the stack we'll pick `a[1]` and `b[1]` and push `stack.push(a, b, i + 1);`

```
    var i = stack.pop();
    b = stack.pop();
    a = stack.pop();
    if (i + 1 < a.length) stack.push(a, b, i + 1);
    a = a[i];
    b = b[i];
```

File: runtime/stdlib.js (hunk 13 above)
Comment (hhugo) [reply runtime/stdlib.js]: No, we pop from it as well.

File: runtime/stdlib.js (hunk 11 above)
Comment (hhugo) [reply runtime/stdlib.js]: Well, it should behave like comparison on floats

File: runtime/stdlib.js (hunk 8 above)
Comment (hhugo) [reply runtime/stdlib.js]: my fear was that using variable in `switch` cases could be slower.
Theses tags are the ocaml tags, with small additions
```
let lazy_tag = 246
let closure_tag = 247
let object_tag = 248
let infix_tag = 249
let forward_tag = 250

let no_scan_tag = 251

let abstract_tag = 251
let string_tag = 252
let double_tag = 253
let double_array_tag = 254
let custom_tag = 255
let final_tag = custom_tag


let int_tag = 1000
let out_of_heap_tag = 1001
let unaligned_tag = 1002
```

Additions are
```
1255 (similar to custom_tag 255)
1256 (for JavaScript objects that have a compare method)  
1247 (similar to closure_tag 1247)
```

File: runtime/stdlib.js (hunk 10 above)
Comment (hhugo) [reply runtime/stdlib.js]: We only return NaN when we don't need a total order.
It correspond to UNORDERED in the the ocaml runtime (see https://github.com/ocaml/ocaml/blob/trunk/runtime/compare.c#L110)

File: runtime/stdlib.js (hunk 7 above)
Comment (hhugo) [reply runtime/stdlib.js]: For instance, `compare nan nan = 0` but `nan <> nan`,`(nan = nan) = false`, `(nan < nan) = false`, `(nan > nan) = false`

File: runtime/stdlib.js (hunk 2 above)
Comment (hhugo) [reply runtime/stdlib.js]: I removed the check above

File: runtime/stdlib.js (hunk 6 above)
Comment (hhugo) [reply runtime/stdlib.js]: renamed

File: lib/js_of_ocaml/import.ml (hunk 4 above)
Comment (hhugo) [reply lib/js_of_ocaml/import.ml]: I believe theses import files will diverge on purpose overtime.

File: runtime/stdlib.js (hunk 10 above)
Comment (TyOverby) [reply runtime/stdlib.js]: If we are expected to return an int, isn't NaN breaking that requirement?

File: runtime/stdlib.js [hunk 14]
```diff
@@ -358,90 +358,205 @@ function caml_floatarray_create(len){
...
+        if (a > b) return 1;
+        if (a != b) {
+          if (!total) return NaN;
+          if (a == a) return 1;
+          if (b == b) return -1;
+        }
+        break;
+      case 246: // Lazy_tag
+      case 1251: // JavaScript Symbol, not ordering.
+        if(a != b) {
```
Comment (TyOverby) [inline runtime/stdlib.js]: should this be `===`?

Comment (TyOverby) [reply runtime/stdlib.js]: or `!==` rather

File: runtime/stdlib.js (hunk 10 above)
Comment (hhugo) [reply runtime/stdlib.js]: caml_compare_val is an helper function for the functions below. We only use its return value directly when `total` is true (in `caml_comare`)
```javascript
//Provides: caml_compare (const, const)
//Requires: caml_compare_val
function caml_compare (a, b) { return caml_compare_val (a, b, true); }
//Provides: caml_equal mutable (const, const)
//Requires: caml_compare_val
function caml_equal (x, y) { return +(caml_compare_val(x,y,false) == 0); }
//Provides: caml_notequal mutable (const, const)
//Requires: caml_compare_val
function caml_notequal (x, y) { return +(caml_compare_val(x,y,false) != 0); }
//Provides: caml_greaterequal mutable (const, const)
//Requires: caml_compare_val
function caml_greaterequal (x, y) { return +(caml_compare_val(x,y,false) >= 0); }
//Provides: caml_greaterthan mutable (const, const)
//Requires: caml_compare_val
function caml_greaterthan (x, y) { return +(caml_compare_val(x,y,false) > 0); }
//Provides: caml_lessequal mutable (const, const)
//Requires: caml_compare_val
function caml_lessequal (x, y) { return +(caml_compare_val(x,y,false) <= 0); }
//Provides: caml_lessthan mutable (const, const)
//Requires: caml_compare_val
function caml_lessthan (x, y) { return +(caml_compare_val(x,y,false) < 0); }
```

File: runtime/stdlib.js (hunk 14 above)
Comment (hhugo) [reply runtime/stdlib.js]: `!==` and `!=` would behave the same for `Symbol`.
But you're right, `!==` convey the intent better

File: lib/js_of_ocaml/import.ml (hunk 4 above)
Comment (hhugo) [reply lib/js_of_ocaml/import.ml]: This diff has been pushed on master already.

File: runtime/stdlib.js [hunk 15]
```diff
@@ -423,8 +424,9 @@ function caml_compare_val (a, b, total) {
             var comp = caml_compare_val_get_custom(a);
             if(comp) {
               var x = comp(a,b,total);
-              if(!(total && x != x) || (x === (x|0) && x != 0))
-                return x
+              if(total && x != x) return 1;
+              if(+x != +x) return +x;
+              if((x | 0) != 0) return (x | 0);
```
Comment (TyOverby) [inline runtime/stdlib.js]: Could you factor the two blocks in this commit out into a function?  I'm terrified of this bithackery getting out of sync.

File: runtime/stdlib.js [hunk 16]
```diff
@@ -358,90 +358,203 @@ function caml_floatarray_create(len){
...
+    // Look like an ocaml block
+    var tag = a[0] | 0;
+    // ignore double_array_tag because we cannot accurately set
+    // this tag when we create an array of float.
+    return (tag == 254)?0:tag
+  }
+  else if (a instanceof String) return 1252; // javascript string, like string_tag (252)
+  else if (typeof a == "string") return 1252; // javascript string, like string_tag (252)
+  else if (a instanceof Number) return 1000; // int_tag (we use it for all numbers)
+  else if (a.caml_custom) return 1255; // like custom_tag (255)
```
Comment (vouillon) [inline runtime/stdlib.js]: @hhugo, this breaks when `a` is `null` or `undefined`

Comment (hhugo) [reply runtime/stdlib.js]: https://github.com/ocsigen/js_of_ocaml/pull/920
